File: impl/go/nix/worked_example.go
package nix

// A real package, through a real overlay, against real nixpkgs.
//
// The eleven conformance intents are derivations nobody would write: no
// dependencies, no stdenv, no composition. They pin the SERIALIZATION. This
// pins the SURFACE, which is where the project's claim actually lives and
// which had the least evidence behind it.
//
// What it exercises that the intents do not: stdenv.mkDerivation, a dependency
// taken from nixpkgs, a value that arrived through an overlay and is
// interpolated into a build phase, and a fixed point tying the overlay's knot.
// Every one of those is something a user would write on their first day and
// none appears in the corpus.
//
// The check is scripts/worked-example.sh: this term and the hand-written
// docs/spec/examples/worked-example.nix must instantiate to the SAME store
// path. Textual similarity is neither expected nor meaningful; the .drv is the
// normal form, for the same reason the round-trip law lives there.

func workedBase() Expr {
	return Attrs(P("greeting", S("hello")), P("version", S("1.0")))
}

// bump raises the version and derives a value from final.
//
// Reading final is what forces the knot to be tied: an overlay that only ever
// looked at prev would evaluate under a plain // and would not test Fix at all.
func bump(final, prev Expr) Expr {
	return Attrs(
		P("version", S("2.0")),
		P("banner", IStr(Sel(prev, "greeting"), " v", Sel(final, "version"))),
	)
}

// WorkedExample builds the term, from a reset name supply so output is stable.
func WorkedExample() Expr {
	Reset()
	return LetIn("pkgs", App(V("import"), SPath("nixpkgs"), Attrs()),
		func(pkgs Expr) Expr {
			return LetIn("final", Fix(workedBase(), bump), func(final Expr) Expr {
				return App(Sel(pkgs, "stdenv", "mkDerivation"), Attrs(
					P("pname", S("img-drv-worked-example")),
					P("version", Sel(final, "version")),
					P("dontUnpack", B(true)),
					P("nativeBuildInputs", L(Sel(pkgs, "hello"))),
					P("buildPhase", IStr("echo ", Sel(final, "banner"), " > out.txt")),
					P("installPhase",
						S("mkdir -p $out/share && cp out.txt $out/share/")),
				))
			})
		})
}

File: impl/python/src/img_drv/nix/worked_example.py
"""A real package, through a real overlay, against real nixpkgs.

The eleven conformance intents are derivations nobody would write: no
dependencies, no stdenv, no composition. They pin the SERIALIZATION. This pins
the SURFACE, which is where the project claim actually lives and which had the
least evidence behind it.

What it exercises that the intents do not: stdenv.mkDerivation, a dependency
taken from nixpkgs, a value that arrived through an overlay and is interpolated
into a build phase, and a fixed point tying the overlay knot. Every one of
those is something a user would write on their first day and none appears in
the corpus.

The check is scripts/worked-example.sh: this term and the hand-written
docs/spec/examples/worked-example.nix must instantiate to the SAME store path.
Textual similarity is neither expected nor meaningful; the .drv is the normal
form, for the same reason the round-trip law lives there.
"""

from __future__ import annotations

from .ast import Expr
from .surface import (
    apply,
    attrs,
    boolean,
    fix,
    istr,
    let_in,
    listing,
    reset,
    select,
    spath,
    str_,
    var,
)

__all__ = ["term"]

BASE = attrs({"greeting": str_("hello"), "version": str_("1.0")})


def bump(final: Expr, prev: Expr) -> Expr:
    """Bump the version, and derive a value from `final`.

    Reading `final` is what forces the knot to be tied: an overlay that only
    ever looked at `prev` would evaluate under a plain `//` and would not
    test `fix` at all.
    """
    return attrs(
        {
            "version": str_("2.0"),
            "banner": istr(
                [select(prev, ["greeting"]), " v", select(final, ["version"])]
            ),
        }
    )


def term() -> Expr:
    """The worked example, from a reset name supply so output is stable."""
    reset()
    return let_in(
        apply(var("import"), [spath("nixpkgs"), attrs({})]),
        lambda pkgs: let_in(
            fix(BASE, bump),
            lambda final: apply(
                select(pkgs, ["stdenv", "mkDerivation"]),
                [
                    attrs(
                        {
                            "pname": str_("img-drv-worked-example"),
                            "version": select(final, ["version"]),
                            "dontUnpack": boolean(True),
                            "nativeBuildInputs": listing(
                                [select(pkgs, ["hello"])]
                            ),
                            "buildPhase": istr(
                                [
                                    "echo ",
                                    select(final, ["banner"]),
                                    " > out.txt",
                                ]
                            ),
                            "installPhase": str_(
                                "mkdir -p $out/share && cp out.txt $out/share/"
                            ),
                        }
                    )
                ],
            ),
            name="final",
        ),
        name="pkgs",
    )
